Fix FlowAggregator e2e test random failures

dreamtalen opened the pull request:
In this commit, we fixed some random failure cases of the FlowAggregator e2e test, which are previously raised by Jianjun in the email thread.

One fix is for the ToExternalFlows test, the reason this test fails randomly is that `wget` is used instead of `iperf -t 12` to generate traffic, so 6 seconds of waiting time is sometimes not enough because the first record theoretically arrives after 5.5 seconds `(exporterActiveFlowExportTimeout+aggregatorActiveFlowRecordTimeout)`. So we changed to it 9 seconds (5.5s plus one more `aggregatorActiveFlowRecordTimeout`) for waiting for flow records received on the IPFIX collector.

Another fix is adding `flowEndReason==3` in the condition of receiving the last flow record. We found in a low possibility, the `flowEndSeconds - flowStartSeconds` of the last flow record is equal to `11` instead of `>= 12`. So we add `flowEndReason==3` in the condition of finding the last flow record. 

Added 2/18:
The third fix is setting a bandwidth limit of iperf command to make the throughput test more stable. We found when running iperf command with unlimited bandwidth by default, the throughput fluctuates greatly (±20%) at the beginning. So we set a limit of 10Mbits/s for it and the throughput test becomes more stable.

I have triggered 5 times of Kind test of Flow Aggregator on GitHub and no failures were found: https://github.com/antrea-io/antrea/runs/5221164243?check_suite_focus=true

Signed-off-by: Yongming Ding <dyongming@vmware.com>

jianjuns reviewed (commented):
I do not see the fix of "flowEndReason==3" in code?

In commit message:
> One fix is for ToExternalFlows test, we gave it 2 more seconds for

test, we gave -> test. We give

dreamtalen commented:
> I do not see the fix of "flowEndReason==3" in code?
> 
> In commit message:
> 
> > One fix is for ToExternalFlows test, we gave it 2 more seconds for
> 
> test, we gave -> test. We give

Thanks Jianjun, I squashed two commits and addressed the message, I changed the code from 
`if exportTime >= flowStartTime+iperfTimeSec {`
to
`if exportTime >= flowStartTime+iperfTimeSec || flowEndReason == 3 {`

heanlan reviewed (dismissed):
Thanks Yongming for adding the fix.

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -852,7 +852,7 @@ func getUint64FieldFromRecord(t *testing.T, record string, field string) uint64
 	var collectorOutput string
 	var recordSlices []string
-	err := wait.PollImmediate(500*time.Millisecond, aggregatorInactiveFlowRecordTimeout, func() (bool, error) {
+	err := wait.PollImmediate(500*time.Millisecond, aggregatorInactiveFlowRecordTimeout+exporterActiveFlowExportTimeout, func() (bool, error) {
```
would be good to have comments in the code to explain the timeout choice when it's not obvious

dreamtalen replied:
Thanks, added.

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -867,8 +867,9 @@ func getCollectorOutput(t *testing.T, srcIP, dstIP, srcPort string, isDstService
+				flowEndReason := int64(getUint64FieldFromRecord(t, record, "flowEndReason"))
 				if strings.Contains(record, src) && strings.Contains(record, dst) && strings.Contains(record, srcPort) {
-					if exportTime >= flowStartTime+iperfTimeSec {
+					if exportTime >= flowStartTime+iperfTimeSec || flowEndReason == 3 {
```
same as above, would be good to have a code comment here to remind the reader of what `flowEndReason== 3` means

dreamtalen replied:
Thanks, added.

dreamtalen commented:
/test-all

heanlan commented on test/e2e/flowaggregator_test.go:
```diff
@@ -129,6 +129,8 @@ const (
 	testEgressRuleName             = "test-egress-rule-name"
 	iperfTimeSec                   = 12
 	protocolIdentifierTCP          = 6
+	// Set a bandwidth limit of iperf command to make the throughput test more stable
```
```suggestion
	// Set target bandwidth(bits/sec) of iPerf traffic to a relatively small value
	// (default unlimited for TCP), to reduce the variances causing by network performance
	// during 12s, and make the throughput test more stable. 
```

antoninbas replied:
+1

dreamtalen replied:
Thanks, addressed.

heanlan commented on test/e2e/flowaggregator_test.go:
```diff
@@ -684,19 +684,20 @@ func checkRecordsForFlows(t *testing.T, data *TestData, srcIP string, dstIP stri
-				t.Logf("Iperf throughput: %.2f Mbits/s, IPFIX record throughput: %.2f Mbits/s", bandwidthInMbps, recBandwidth)
-				assert.InDeltaf(recBandwidth, bandwidthInMbps, bandwidthInMbps*0.15, "Difference between Iperf bandwidth and IPFIX record bandwidth should be lower than 15%%")
+				t.Logf("Throughput check on record with flowEndSeconds-flowStartSeconds: %v, Iperf throughput: %.2f Mbits/s, IPFIX record throughput: %.2f Mbits/s", exportTime-flowStartTime, bandwidthInMbps, recBandwidth)
+				assert.InDeltaf(recBandwidth, bandwidthInMbps, bandwidthInMbps*0.15, "Difference between Iperf bandwidth and IPFIX record bandwidth should be lower than 15%%, record: %s", record)
```
```suggestion
				assert.InDeltaf(recBandwidth, bandwidthInMbps, bandwidthInMbps*0.15, "Difference between Iperf bandwidth and IPFIX record bandwidth should be lower than 15%, record: %s", record)
```

heanlan reviewed (commented):
A few nits.

dreamtalen commented:
@antoninbas @jianjuns Hi Antonin and Jianjun, may I have a second round of reviews from you? We added another fix for the throughput test. Since this PR could be beneficial for merging other PRs by reducing the test failure, we would like to merge it sooner, thanks!

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -129,6 +129,8 @@ const (
 	testEgressRuleName             = "test-egress-rule-name"
 	iperfTimeSec                   = 12
 	protocolIdentifierTCP          = 6
+	// Set a bandwidth limit of iperf command to make the throughput test more stable
```
this is a good idea

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -625,10 +627,8 @@ func checkRecordsForFlows(t *testing.T, data *TestData, srcIP string, dstIP stri
-		bandwidthInMbps = bandwidthInFloat * float64(1024)
 	} else {
-		t.Fatalf("Unit of the traffic bandwidth reported by iperf should either be Mbits or Gbits, failing the test.")
+		t.Fatalf("Unit of the traffic bandwidth reported by iperf should either be Mbits, failing the test.")
```
```suggestion
		t.Fatalf("Unit of the traffic bandwidth reported by iperf should be Mbits.")
```

dreamtalen replied:
Thanks, addressed.

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -684,19 +684,20 @@ func checkRecordsForFlows(t *testing.T, data *TestData, srcIP string, dstIP stri
-				// Check bandwidth with the field "throughput" except for the last record,
-				// as its throughput is significantly lower than the average Iperf throughput.
-				if exportTime < flowStartTime+iperfTimeSec {
+				if exportTime >= flowStartTime+iperfTimeSec || flowEndReason == 3 {
```
maybe add the same comment as below for `flowEndReason == 3`

dreamtalen replied:
Thanks, added.

antoninbas commented on test/e2e/flowaggregator_test.go:
```diff
@@ -852,7 +853,9 @@ func getUint64FieldFromRecord(t *testing.T, record string, field string) uint64
 	var recordSlices []string
-	err := wait.PollImmediate(500*time.Millisecond, aggregatorInactiveFlowRecordTimeout, func() (bool, error) {
+	// In the ToExternalFlows test, flow record will arrive 5.5s (exporterActiveFlowExportTimeout+aggregatorActiveFlowRecordTimeout) after executing wget command
+	// We set the timeout to 8s (exporterActiveFlowExportTimeout+aggregatorInactiveFlowRecordTimeout) to make the ToExternalFlows test more stable
```
It seems reasonable to me to wait for 8s, which is slightly larger than the expected delay (5.5s). However, do we choose this value _because_ it is slightly larger than 5.5s or doe choose it because it is `exporterActiveFlowExportTimeout+aggregatorInactiveFlowRecordTimeout`? In other words, if `aggregatorInactiveFlowRecordTimeout` was larger (e.g. 20s), would be still be using `exporterActiveFlowExportTimeout+aggregatorInactiveFlowRecordTimeout`?

antoninbas replied:
That sounds good, thanks

dreamtalen replied:
Cool, updated.

dreamtalen commented:
/test-all

dreamtalen commented:
/test-integration

tnqn reviewed (approved):
Thanks for the fixes!

tnqn commented:
/test-e2e
/test-integration
